cmd/kubeadm/app/cmd/upgrade/plan.go

```diff
@@ -92,85 +156,63 @@ func runPlan(flags *planFlags, userVersion string) error {
-			fmt.Fprintf(tabw, "Etcd\t%s\t%s\n", upgrade.Before.EtcdVersion, upgrade.After.EtcdVersion)
-
-			// We should flush the writer here at this stage; as the columns will now be of the right size, adjusted to the above content
-			tabw.Flush()
-			fmt.Fprintln(w, "")
+			printer.Fprintf(writer, "External components that should be upgraded manually before you upgrade the control plane with 'kubeadm upgrade apply':\n")
+			printer.PrintObj(newUpgradePlanItem("Etcd", upgrade.Before.EtcdVersion, upgrade.After.EtcdVersion), writer)
+			printer.Flush()
+			printer.Fprintf(writer, "\n")
 		}
 
 		if upgrade.CanUpgradeKubelets() {
-			fmt.Fprintln(w, "Components that must be upgraded manually after you have upgraded the control plane with 'kubeadm upgrade apply':")
-			fmt.Fprintln(tabw, "COMPONENT\tCURRENT\tAVAILABLE")
-			firstPrinted := false
+			printer.Fprintf(writer, "Components that must be upgraded manually after you have upgraded the control plane with 'kubeadm upgrade apply':\n")
 
-			// The map is of the form <old-version>:<node-count>. Here all the keys are put into a slice and sorted
-			// in order to always get the right order. Then the map value is extracted separately
+			name := "Kubelet"
 			for _, oldVersion := range sortedSliceFromStringIntMap(upgrade.Before.KubeletVersions) {
 				nodeCount := upgrade.Before.KubeletVersions[oldVersion]
-				if !firstPrinted {
-					// Output the Kubelet header only on the first version pair
-					fmt.Fprintf(tabw, "Kubelet\t%d x %s\t%s\n", nodeCount, oldVersion, upgrade.After.KubeVersion)
-					firstPrinted = true
-					continue
+				item := newUpgradePlanItem(name, fmt.Sprintf("%d x %s", nodeCount, oldVersion), upgrade.After.KubeVersion)
+				err = printer.PrintObj(item, writer)
+				if err != nil {
+					return errors.Wrapf(err, "can't print upgrade plan item %v", item)
 				}
-				fmt.Fprintf(tabw, "\t%d x %s\t%s\n", nodeCount, oldVersion, upgrade.After.KubeVersion)
+				name = ""
 			}
-			// We should flush the writer here at this stage; as the columns will now be of the right size, adjusted to the above content
-			tabw.Flush()
-			fmt.Fprintln(w, "")
 		}
 
-		fmt.Fprintf(w, "Upgrade to the latest %s:\n", upgrade.Description)
-		fmt.Fprintln(w, "")
-		fmt.Fprintln(tabw, "COMPONENT\tCURRENT\tAVAILABLE")
-		fmt.Fprintf(tabw, "API Server\t%s\t%s\n", upgrade.Before.KubeVersion, upgrade.After.KubeVersion)
-		fmt.Fprintf(tabw, "Controller Manager\t%s\t%s\n", upgrade.Before.KubeVersion, upgrade.After.KubeVersion)
-		fmt.Fprintf(tabw, "Scheduler\t%s\t%s\n", upgrade.Before.KubeVersion, upgrade.After.KubeVersion)
-		fmt.Fprintf(tabw, "Kube Proxy\t%s\t%s\n", upgrade.Before.KubeVersion, upgrade.After.KubeVersion)
+		printer.Flush()
+		printer.Fprintf(writer, "\nUpgrade to the latest %s:\n\n", upgrade.Description)
+
+		items := []*outputapiv1alpha1.UpgradePlanItem{
+			newUpgradePlanItem("API Server", upgrade.Before.KubeVersion, upgrade.After.KubeVersion),
```

Comment:
This is interesting. Ideally, in the machine output one would like a more machine friendly component name here (`constants.KubeAPIServer`). However, that's not ideal for end users (text output), where we would want to keep `API Server`.
We can try to have both somehow and, depending on the output type, dump the preferred one?

Reply:
Yes, we can, but it would look quite ugly. We'll have to wrap all similar places into something like 
```
if out  == output.TextOutput { 
    name = "API Servier"
} else {
    name = constants.KubeaAPIServer
} 
``` everywhere we want to use this.

Any better idea how to do it?

Reply:
I'll think of some Go magic to hide this. You are right, that ```if...else``` everywhere is barely readable and ugly.

Reply:
using kube-apiserver for both is fine and i've seen that in other places in k/k.

Reply:
@neolit123 thank you for the review. this should be done now.